zerver/views/message_fetch.py

```diff
@@ -306,6 +307,35 @@ def by_streams(self, query: Select, operand: str, maybe_negate: ConditionTransfo
             recipient_queryset = get_public_streams_queryset(self.realm)
         elif operand == "web-public":
             recipient_queryset = get_web_public_streams_queryset(self.realm)
+        elif operand == "subscribed":
+            subscribed_streams_with_full_history = (
+                get_subscribed_streams_for_user_by_history_public_to_subscribers(
+                    self.realm, self.user_profile, True
+                )
+            )
+            subscribed_streams_with_limited_history = (
+                get_subscribed_streams_for_user_by_history_public_to_subscribers(
+                    self.realm, self.user_profile, False
+                )
+            )
+            search_recipient_ids = subscribed_streams_with_limited_history.values_list(
+                "recipient_id", flat=True
+            ).order_by("id")
+            full_history_recipient_ids = subscribed_streams_with_full_history.values_list(
+                "recipient_id", flat=True
+            ).order_by("id")
+            subscribed_recipient_cond = column("recipient_id", Integer).in_(search_recipient_ids)
+            full_history_recipient_cond = column("recipient_id", Integer).in_(
+                full_history_recipient_ids
+            )
+
+            user_message_query = join_with_usermessage(self.user_profile, query)
```

Comment:
I worry this strategy of doing a late JOIN is not going to work for queries like `is:starred streams:subscribed`.  

Certainly the architecture of this module does not expect we'll need to do a join inside an individual narrow operator rule, but I see why you need to break that invariant to implement this correctly.  

Probably worth fixing the other issues (E.g. I think the basic `join_with_usermessage` extraction is a good refactor in any case) and then having a discussion about how to resolve this thorny issue.